### PR TITLE
[gh-17] [API] As a user, I can sign-up using email and password

### LINKED ISSUE #17: [API] As a user, I can sign-up using email and password
## Why

User needs to use the sign-up function as an API for calling from other client-side like mobile.

## Who Benefits?

The user who wants to use our application functions from other client-side application

### PR DESCRIPTION
**Pull request by nvminhtue**
https://github.com/nvminhtue/elixir_internal_certificate/issues/17

## What happened

Implement API part of registration new account
 
## Insight

- Add config of JSONAPI
- Config and handle router to API V1 `user_registration`

## Proof Of Work
Successful case

image

Failed cases
- Existing email

image

- Blank email/password

image

### CODE REVIEW

File: lib/elixir_internal_certificate_web/controllers/api/v1/user_registration_controller.ex [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  alias ElixirInternalCertificateWeb.Api.V1.UserRegistrationView
+
+  def create(conn, params) do
+    with changeset <- Accounts.change_user_registration(%User{}, params),
+         {:ok, user} <- Accounts.register_user(changeset.changes) do
+      conn
+      |> put_view(UserRegistrationView)
+      |> render("show.json", %{data: user})
+    else
+      {:error, %Ecto.Changeset{} = _changeset} ->
```
Comment (andyduong1920) [inline lib/elixir_internal_certificate_web/controllers/api/v1/user_registration_controller.ex]: should we move this error to `fallback controller` @nvminhtue 🤔

File: lib/elixir_internal_certificate_web/controllers/api/v1/user_registration_controller.ex [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        conn
+        |> put_view(ErrorView)
+        |> put_status(:bad_request)
+        |> render("error.json", %{code: :unprocessable_entity, message: "Invalid email or password"})
```
Comment (andyduong1920) [inline lib/elixir_internal_certificate_web/controllers/api/v1/user_registration_controller.ex]: The `invalid email or password` error message doesn't make much sense on creation, on creation, it's better to let the user know what is the actual error for email email format is invalid or password is invalid, it should contains upercawe, lowercase, number, ... for example...

File: lib/elixir_internal_certificate_web/controllers/api/v1/user_registration_controller.ex [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  alias ElixirInternalCertificate.Account.Accounts
+  alias ElixirInternalCertificate.Account.Schemas.User
+  alias ElixirInternalCertificateWeb.Api.ErrorView
+  alias ElixirInternalCertificateWeb.Api.V1.UserRegistrationView
+
+  def create(conn, params) do
+    with changeset <- Accounts.change_user_registration(%User{}, params),
+         {:ok, user} <- Accounts.register_user(changeset.changes) do
+      conn
+      |> put_view(UserRegistrationView)
```
Comment (andyduong1920) [inline lib/elixir_internal_certificate_web/controllers/api/v1/user_registration_controller.ex]: if we add the `UserRegistrationView` in `V1` module, we no need to use the `put_view` here

```
V1.UserRegistrationView
```

File: lib/elixir_internal_certificate_web/controllers/api/v1/user_registration_controller.ex [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  use ElixirInternalCertificateWeb, :controller
+
+  alias ElixirInternalCertificate.Account.Accounts
+  alias ElixirInternalCertificate.Account.Schemas.User
+  alias ElixirInternalCertificateWeb.Api.ErrorView
+  alias ElixirInternalCertificateWeb.Api.V1.UserRegistrationView
+
+  def create(conn, params) do
+    with changeset <- Accounts.change_user_registration(%User{}, params),
+         {:ok, user} <- Accounts.register_user(changeset.changes) do
```
Comment (andyduong1920) [inline lib/elixir_internal_certificate_web/controllers/api/v1/user_registration_controller.ex]: how about

```
validated_params <-
 {:ok, user} <- Accounts.register_user(validated_params) do
```

File: lib/elixir_internal_certificate_web/views/api/v1/user_registration_view.ex [hunk 5]
```diff
@@ -0,0 +1,10 @@
+defmodule ElixirInternalCertificateWeb.Api.V1.UserRegistrationView do
+  use JSONAPI.View, type: "user_auth"
+
+  def fields do
+    [
+      :email,
+      :password
```
Comment (andyduong1920) [inline lib/elixir_internal_certificate_web/views/api/v1/user_registration_view.ex]: Are we return the password to the API response, might I know the reason for that @nvminhtue 🤔

File: test/elixir_internal_certificate_web/controllers/api/v1/user_registration_controller_test.exs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        post(conn, Routes.api_user_registration_path(conn, :create), %{
+          email: "existing@mail.com",
+          password: valid_user_password()
+        })
+
+      assert json_response(conn, 400) == %{
+               "errors" => [
+                 %{
+                   "code" => "unprocessable_entity",
+                   "message" => "Invalid email or password"
```
Comment (andyduong1920) [inline test/elixir_internal_certificate_web/controllers/api/v1/user_registration_controller_test.exs]: As a normal user, they won't know what's wrong with their email or password here :P

File: test/elixir_internal_certificate_web/controllers/api/v1/user_registration_controller_test.exs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        post(conn, Routes.api_user_registration_path(conn, :create), %{
+          email: "",
+          password: ""
+        })
+
+      assert json_response(conn, 400) == %{
+               "errors" => [
+                 %{
+                   "code" => "unprocessable_entity",
+                   "message" => "Invalid email or password"
```
Comment (andyduong1920) [inline test/elixir_internal_certificate_web/controllers/api/v1/user_registration_controller_test.exs]: Should we change the message to `password can't be blank, email can't be blank`

File: lib/elixir_internal_certificate_web/controllers/api/v1/user_registration_controller.ex (hunk 1 above)
Comment (nvminhtue) [reply lib/elixir_internal_certificate_web/controllers/api/v1/user_registration_controller.ex]: Updated in [4fc3f20](https://github.com/nvminhtue/elixir_internal_certificate/pull/42/commits/4fc3f20c5b2b4369b4ca39bf951d756d7bc7f928)

File: lib/elixir_internal_certificate_web/controllers/api/v1/user_registration_controller.ex [hunk 8]
```diff
@@ -0,0 +1,17 @@
+defmodule ElixirInternalCertificateWeb.Api.V1.UserRegistrationController do
+  use ElixirInternalCertificateWeb, :controller
+
+  alias ElixirInternalCertificate.Account.Accounts
+  alias ElixirInternalCertificate.Account.Schemas.User
+
+  def create(conn, params) do
+    with %{changes: validated_params} = _changeset <-
+           Accounts.change_user_registration(%User{}, params),
```
Comment (andyduong1920) [inline lib/elixir_internal_certificate_web/controllers/api/v1/user_registration_controller.ex]: we can remove `%{changes: validated_params}` I think

File: lib/elixir_internal_certificate_web/controllers/api/v1/user_registration_controller.ex [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+         {:ok, user} <- Accounts.register_user(validated_params) do
+      render(conn, "show.json", %{data: user})
+    else
```
Comment (andyduong1920) [inline lib/elixir_internal_certificate_web/controllers/api/v1/user_registration_controller.ex]: we can remove this as we had it in the fallback

```
   else
      error ->
        error
```

**Review by andyduong1920 — APPROVED** [review]

File: lib/elixir_internal_certificate_web/controllers/api/v1/user_registration_controller.ex (hunk 9 above)
Comment (nvminhtue) [reply lib/elixir_internal_certificate_web/controllers/api/v1/user_registration_controller.ex]: Removed both in [fbc2ced](https://github.com/nvminhtue/elixir_internal_certificate/pull/42/commits/fbc2ced7ecddd38be8c84a18f6c6ca0f93a46f99)

**Review by longnd — APPROVED** [review]

**Pull request merged**